stripe/stripe_object.py

```diff
@@ -78,20 +111,22 @@ def __init__(
     def last_response(self) -> Optional[StripeResponse]:
         return self._last_response
 
-    def update(self, update_dict: Dict[str, Any]):
+    # StripeObject inherits from `dict` which has an update method, and this doesn't quite match
+    # the full signature of the update method in MutableMapping. But we ignore.
+    def update(self, update_dict: SupportsKeysAndGetItem[str, Any]) -> None:  # type: ignore[override]
         for k in update_dict:
             self._unsaved_values.add(k)
 
         return super(StripeObject, self).update(update_dict)
 
-    def __setattr__(self, k, v):
-        if k[0] == "_" or k in self.__dict__:
-            return super(StripeObject, self).__setattr__(k, v)
+    if not TYPE_CHECKING:
 
-        self[k] = v
-        return None
+        def __setattr__(self, k, v):
```

Comment:
I thought having these under `TYPE_CHECKING` didn't matter for type checking

Reply:
It does. I should have called this out in the PR description.

This will cause

`customer.this_property_does_not_exist = "foo".

to begin producing a type error. I think when we were originally working on the magic getter/setter we didn't think this would matter much, but this does matter for the "mutate and call `.save()`" use case, and we got some feedback from internal users about this (see #developer-sdks).

I'm not wedded to this but I don't see too much downside.